### PR TITLE
Issues/101 - Include CosmoDC2 population comparison

### LINKED ISSUE #101: Add CosmoDC2 comparison to Run 2.1i dr1b outputs.
Compare `cosmoDC` GCR catalog to Run 2.1 dr1b outputs
in `validation/validate_dc2_run2.1i_object_table.ipynb`

This will be focused on compare distributions.

No matching-based comparison.  Those will be done in a separate issue.  See, e.g., #90

### PR DESCRIPTION
**Pull request by wmwv**
Compare number density and areal density to CosmoDC2

Does not do matching.  Just compares populations things like RA, Dec distribution and number density with magnitude.

Rendered version available at
https://github.com/LSSTDESC/DC2-analysis/blob/rendered/validation/validate_dc2_run2.1i_object_table.ipynb

If that doesn't load for you through the GitHub render, clone the `DC2-analysis` repo and check out the `rendered` branch locally.

fix #101

### CODE REVIEW

File: validation/validate_dc2_run2.1i_object_table.ipynb [hunk 1]
```diff
@@ -28,7 +28,13 @@
     "  * 78 million objects  \n",
     "  * 48 million objects with i-band SNR > 5\n",
     "\n",
-    "Loading the entire catalog takes ~15 minutes, depending on memory load on the JupyterHub node.  We've thus entered an era where it will be common to only look at subsamples of the full DC2 datasets, whether that's considering just one tract, or taking a subsample of the full catalog."
+    "Loading the entire catalog takes ~15 minutes, depending on memory load on the JupyterHub node.  We've thus entered an era where it will be common to only look at subsamples of the full DC2 datasets, whether that's considering just one tract, or taking a subsample of the full catalog.\n",
```
Comment (kadrlica) [inline validation/validate_dc2_run2.1i_object_table.ipynb]: Does this memory get cached? I'm wondering if the load time is much quicker the second time you load the catalog. As a DES person, we are working with much larger catalogs and I still analyze the full data set, so I'm not sure that 78M objects is really the "beginning of an era".

Comment (kadrlica) [reply validation/validate_dc2_run2.1i_object_table.ipynb]: The `math` import seems a bit unnecessary with the `numpy` import.

Comment (kadrlica) [reply validation/validate_dc2_run2.1i_object_table.ipynb]: and I think I see the answer to my caching question in cell 8 `'use_cache': False`.

File: validation/validate_dc2_run2.1i_object_table.ipynb [hunk 2]
```diff
@@ -224,6 +247,134 @@
...
+   "outputs": [],
+   "source": [
+    "# Establish some basic cuts intended to yield a galaxy sample with reasonable flux measurements.\n",
+    "basic_cuts = [\n",
+    "    GCRQuery('clean'), # The source has no flagged pixels (interpolated, saturated, edge, clipped...) \n",
+    "                       # and was not skipped by the deblender\n",
+    "    GCRQuery('xy_flag == 0'),                                      # Flag for bad centroid measurement\n",
+    "]\n",
+    "\n",
+    "# Define cuts on galaxy properties: currently we simply make a sample limited at i<25.\n",
```
Comment (kadrlica) [inline validation/validate_dc2_run2.1i_object_table.ipynb]: I didn't see where any star/galaxy separation occurred. Is this really cuts on "object" properties?

File: validation/validate_dc2_run2.1i_object_table.ipynb [hunk 3]
```diff
@@ -224,6 +247,134 @@
...
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "hsc = pd.DataFrame(hsc_cat.get_quantities(hsc_columns,\n",
+    "                                          filters=basic_cuts + properties_cuts))\n",
+    "\n",
+    "# We could of course done this as part of a GCR query.\n",
+    "# But using a query in the DataFrame is how we did the `stars`, `galaxies` selection above, so we'll do it the same way here.\n",
```
Comment (kadrlica) [inline validation/validate_dc2_run2.1i_object_table.ipynb]: Where did we do this above?

File: validation/validate_dc2_run2.1i_object_table.ipynb [hunk 4]
```diff
@@ -280,19 +431,20 @@
...
+    "        ax.add_patch(wfd_region)\n",
+    "        ax.add_patch(ddf_region)\n",
+    "\n",
+    "        max_delta_ra = dc2_run2x_wfd[3][0] - dc2_run2x_wfd[2][0]\n",
+    "        delta_dec = dc2_run2x_wfd[1][1] - dc2_run2x_wfd[3][1]\n",
+    "        grow_buffer = 0.05\n",
+    "        ax.set_xlim(dc2_run2x_wfd[3][0] + max_delta_ra * grow_buffer,\n",
+    "                    dc2_run2x_wfd[2][0] - max_delta_ra * grow_buffer)\n",
+    "        ax.set_ylim(dc2_run2x_wfd[3][1] - delta_dec * grow_buffer,\n",
+    "                    dc2_run2x_wfd[1][1] + delta_dec * grow_buffer)"
```
Comment (kadrlica) [inline validation/validate_dc2_run2.1i_object_table.ipynb]: The text says: "We're missing a bit of the South-West (lower-left, "North Up, East Left") corner." The gap that I see is in the lower right (but maybe I'm not looking at the right gap?

Also, you aren't a fan of matplotlib's [`invert_xaxis`](https://matplotlib.org/3.1.0/api/_as_gen/matplotlib.axes.Axes.invert_xaxis.html)?

Comment (kadrlica) [reply validation/validate_dc2_run2.1i_object_table.ipynb]: For the `inside_trapezoid` function you might consider making a path and then using matplotlib [Path.contains_points](https://matplotlib.org/3.1.1/api/path_api.html#matplotlib.path.Path.contains_points). It won't be quite as fast, but more flexible in shape. I guess this is a question of what you want to teach.

File: validation/validate_dc2_run2.1i_object_table.ipynb (hunk 3 above)
Comment (kadrlica) [reply validation/validate_dc2_run2.1i_object_table.ipynb]: I think we do this below in cell 32.

File: validation/validate_dc2_run2.1i_object_table.ipynb (hunk 1 above)
Comment (yymao) [reply validation/validate_dc2_run2.1i_object_table.ipynb]: The load time statement here is outdated I think, with parquet files the time should be much shorter, no? And the `use_cache` option is no longer relevant for parquet files.

Comment (wmwv) [reply validation/validate_dc2_run2.1i_object_table.ipynb]: Oh, I meant local "era", I wasn't trying to be more grandiose.

Comment (wmwv) [reply validation/validate_dc2_run2.1i_object_table.ipynb]: `math.log10` on scalars is several times faster than `numpy.log10`
I haven't explicited performance tested `math.ceil`.  Nor have I explictly tested the performance of the lambda written with `math.ceil(...math.log10(...))` vs. `np.ceil(...np.log10(...))`.  It's quite possible that the disk I/O time is dominant.

But that's why I used it.

Comment (wmwv) [reply validation/validate_dc2_run2.1i_object_table.ipynb]: Our effective storage I/O performance on NERSC is substantially worse than is possible from a direct disk on a computer.  We haven't fully investigated, or worked with NERSC to investigate.  Performance to/from scratch is better, as is to/from the burst buffer, but those aren't for permanent storage.

File: validation/validate_dc2_run2.1i_object_table.ipynb (hunk 3 above)
Comment (wmwv) [reply validation/validate_dc2_run2.1i_object_table.ipynb]: Ah, it used to be above.  I've removed any mention of relative direction.

File: validation/validate_dc2_run2.1i_object_table.ipynb (hunk 4 above)
Comment (wmwv) [reply validation/validate_dc2_run2.1i_object_table.ipynb]: Agreed that in my overall practice I should use `invert_[xy]axis` more often.  In this particular Notebook I'm specifically controlling the limits for each of the `set_[xy]lim` calls.  E.g., in the above, I'm explicitly adding a buffer around the x limits.

There's one `plt.ylim(...)` that's used to flip.  I'll consider changing it.

File: validation/validate_dc2_run2.1i_object_table.ipynb (hunk 1 above)
Comment (wmwv) [reply validation/validate_dc2_run2.1i_object_table.ipynb]: @yymao Yes, the Parquet files are better.  But it still takes 15 minutes to load the entire catalog.

**Pull request merged**